### PR TITLE
cluster: support multiple datacenters

### PR DESCRIPTION
**Pull request by amoskong**
Introduced a new parameter 'n_db_nodes_dcs', it's number list of database
nodes in multiple datecenter.

The multi-dc implement is using one gce region right now, we can extend it to multiple regions in future.

Reference: http://docs.scylladb.com/procedures/create_cluster_multidc/

### CODE REVIEW

File: sdcm/cluster.py [hunk 1]
```diff
@@ -1362,7 +1364,13 @@ def __init__(self, cluster_uuid=None, cluster_prefix='cluster',
         self.nodes = []
         self.nemesis = []
         self.params = params
-        self.add_nodes(n_nodes)
+        self.datacenter = []
+        if type(n_nodes) == list:
```
Comment (larisau) [inline sdcm/cluster.py]: why do not you use isinstance(n_nodes, list)?

File: sdcm/cluster.py [hunk 2]
```diff
@@ -1413,6 +1421,23 @@ def get_node_database_errors(self):
                 errors.append({node.name: node_errors})
         return errors
 
+    def nodes_in_other_dc(self, node, node_list):
+        ret = []
+        for n in node_list:
+            if n.dc_idx != node.dc_idx:
+                ret.append(n)
+        return ret
```
Comment (larisau) [inline sdcm/cluster.py]: can do it in one line:
return [n for n in node_list if n.dc_idx != node.dc_idx]

File: sdcm/cluster.py [hunk 3]
```diff
@@ -2968,6 +2999,21 @@ def _node_setup(self, node, seed_address):
...
+        rackdc_dst_path = os.path.join(tempfile.mkdtemp(prefix='scylla-longevity'),
+                                       'cassandra-rackdc.properties')
+        node.remoter.receive_files(src='/etc/scylla/cassandra-rackdc.properties',
+                                   dst=rackdc_dst_path)
+        with open(rackdc_dst_path, 'r') as f:
+            rackdc_contents = f.read()
+        rackdc_contents += "\ndc=DataCenter%s\n" % (node.dc_idx + 1)
+        rackdc_contents += "\nrack=RACK1\n"
+        with open(rackdc_dst_path, 'w') as f:
+            f.write(rackdc_contents)
```
Comment (larisau) [inline sdcm/cluster.py]: why do not open the file once only - for append, so no need to load data to memory?

File: sdcm/tester.py [hunk 4]
```diff
@@ -262,6 +262,11 @@ def get_cluster_gce(self, loader_info, db_info, monitor_info):
             loader_info['n_local_ssd'] = self.params.get('gce_n_local_ssd_disk_loader')
         if db_info['n_nodes'] is None:
             db_info['n_nodes'] = self.params.get('n_db_nodes')
+        if db_info['n_nodes_dcs'] is None:
+            db_info['n_nodes_dcs'] = self.params.get('n_db_nodes_dcs')
+            db_info['n_nodes'] = []
+            for i in self.params.get('n_db_nodes_dcs').split():
```
Comment (larisau) [inline sdcm/tester.py]: db_info['n_nodes_dcs'] can be used here instead of getting the param twice, and may be It's worth checking it's not None

File: sdcm/cluster.py (hunk 2 above)
Comment (amoskong) [inline sdcm/cluster.py]: It's better, thanks.

File: sdcm/cluster.py (hunk 1 above)
Comment (amoskong) [inline sdcm/cluster.py]: It just works, I will use isinstance()

File: sdcm/tester.py (hunk 4 above)
Comment (amoskong) [inline sdcm/tester.py]: Good catch, I will not use this new parameter.

File: sdcm/cluster.py (hunk 3 above)
Comment (amoskong) [inline sdcm/cluster.py]: There is not effect config in cassandra-rackdc.properties before setup, current simple setup can be done by append.

I will update the remote cassandra-rackdc.properties by:
node.remoter.run("sudo -c 'echo \"\ndc=xxx\nrack=RACK1\" >>  /etc/scylla/cassandra-rackdc.properties'")

File: data_dir/scylla.yaml [hunk 5]
```diff
@@ -6,8 +6,8 @@ test_duration: 60
 stress_cmd: cassandra-stress write cl=QUORUM duration=60m -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate threads=1000 -pop seq=1..10000000
 # tcpdump to try to capture nodetool API
 tcpdump: false
-# Number of database nodes
-n_db_nodes: 6
+# Number list of database nodes in multiple date centers. Legacy type (int): Number of database nodes
+n_db_nodes: '6'
```
Comment (roydahan) [inline data_dir/scylla.yaml]: Do we support asymmetric number of nodes per DC?

File: sdcm/cluster.py [hunk 6]
```diff
@@ -1413,6 +1484,16 @@ def get_node_database_errors(self):
                 errors.append({node.name: node_errors})
         return errors
 
+    def set_tc(self, node, dst_nodes):
+        node.remoter.run("sudo modprobe sch_netem")
+        node.remoter.run("sudo tc qdisc del dev eth0 root", ignore_status=True)
+        node.remoter.run("sudo tc qdisc add dev eth0 handle 1: root prio")
+        for dst in dst_nodes:
+            node.remoter.run("sudo tc filter add dev eth0 protocol ip parent 1:0 prio 1 u32 match ip dst {} flowid 2:1".format(dst.private_ip_address))
+        node.remoter.run("sudo tc qdisc add dev eth0 parent 1:1 handle 2:1 netem delay 100ms 20ms 25% reorder 5% 25% loss random 5% 25%")
```
Comment (roydahan) [inline sdcm/cluster.py]: let's have the delays and loss percentage as parameters.
In case we would like to run the tests with other scenarios.
We can dedicate a section for 'tc' in the yaml file.

Comment (roydahan) [inline sdcm/cluster.py]: Few more things to control the tc:
- Use tc or not (Enabled / Disabled)
- Choose the tc effect (delay / reorder / loss)
- Enable/Disable per host or for ALL or for some of the hosts e.g. (ONE, SOME, ALL)
- How long do we want the tc to be enabled (timer? entire test?)
- Network device ('eth0' may be the default but need to support others).

(These improvements can be in a different PR)

File: data_dir/scylla.yaml [hunk 7]
```diff
@@ -6,8 +6,8 @@ test_duration: 60
 stress_cmd: cassandra-stress write cl=QUORUM duration=60m -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate threads=1000 -pop seq=1..10000000
 # tcpdump to try to capture nodetool API
 tcpdump: false
-# Number of database nodes
-n_db_nodes: 6
+# Number list of database nodes in multiple date centers. Legacy type (int): Number of database nodes
```
Comment (amoskong) [inline data_dir/scylla.yaml]: The parameter is a list string. each iteam is the nodes number of the single dc.

n_db_nodes: '1 2 3'  ===> dc1 has 1 node, dc2 has 2 nodes, dc3 has 3 nodes.

The parameter also indicates the number of dcs.

File: data_dir/scylla.yaml (hunk 5 above)
Comment (amoskong) [inline data_dir/scylla.yaml]: Yes.

File: data_dir/scylla.yaml [hunk 8]
```diff
@@ -118,6 +118,20 @@ backends: !mux
             ami_db_cassandra_user: 'ubuntu'
             ami_id_monitor: 'ami-79d3f06e'
             ami_monitor_user: 'centos'
+        #multiple datacenters
+        us_east_1_and_us_west_2:
```
Comment (amoskong) [inline data_dir/scylla.yaml]: how about multi_aws_dcs? or aws_regions

File: data_dir/scylla.yaml (hunk 7 above)
Comment (roydahan) [inline data_dir/scylla.yaml]: ok, please mention it in the comment above the parameter.

File: data_dir/scylla.yaml (hunk 8 above)
Comment (roydahan) [inline data_dir/scylla.yaml]: Can we have something like?
multi-DC: true
multi-regions: true 
number-of-regions: 2

In this case it will select the first 2 regions in the selected backend_mux for the run.

Comment (amoskong) [inline data_dir/scylla.yaml]: The config parse is done by multiplex plugin in Avocado, it passes single variant to each subtest. I didn't find a way to implement your suggestion.

Comment (roydahan) [inline data_dir/scylla.yaml]: ok, so we can use mux called multi-region-aws and multi-region-gce.

**Comment by roydahan** [comment]
@abvgedeika please help with review.
@amoskong, let's address all ASAP and merge.

**Comment by abvgedeika** [comment]
I guess the implementation should be okay for now. @amoskong  please fix CI issues

**Comment by roydahan** [comment]
Are we merging this PR?
